### PR TITLE
Add line-range pinning for precise staleness detection

### LINKED ISSUE #170: Line-range pinning for precise staleness detection
Extension of #169 (SHA pinning). Instead of flagging a belief as potentially stale whenever any line in the source file changes, store the specific line range the belief depends on.

## Current state (#169)

File SHA changed -> belief potentially stale. Catches everything but has false positives (docstring typo triggers stale flag on unrelated belief).

## Proposed enhancement

Store line span alongside file SHA: e.g. two_phase_commit.py:89-93. Staleness check becomes: did the diff between old SHA and new SHA touch lines 89-93? Still pure mechanical diffing, no LLM, but far fewer false alarms.

Many beliefs already cite specific line numbers in their source entries (recover() at 89-93, lock acquisition at 30-31). The metadata exists -- it just needs to be structured and checkable.

## Origin

Opus 4.8 review of expert.ftl2.com -- proposed as precision upgrade after SHA pinning.

### PR DESCRIPTION
**Pull request by benthomasson**
## Summary

Closes #170. Extends #169 SHA pinning so beliefs can be pinned to specific source line ranges. check-stale --git now only flags stale when the diff between pinned SHA and HEAD touches the pinned lines, not the entire file.

- New metadata field pinned_lines (e.g., 89-93) stored alongside pinned_sha
- New parse_diff_hunks() parses unified diff @@ headers (uses -U0 for exact ranges)
- New file_lines_changed_since() checks committed + uncommitted diffs for line-range overlap
- New pin_lines() sets line range and auto-pins SHA if not already set
- New CLI command: reasons pin-lines NODE START END
- Enhanced reasons show to display pinned SHA and line range
- 21 new tests covering hunk parsing, line-range detection, pin-lines, and check-stale integration

## Test plan

- [x] parse_diff_hunks unit tests (6 tests)
- [x] file_lines_changed_since integration tests (5 tests)
- [x] pin_lines tests (6 tests)
- [x] check_stale line-range integration (4 tests)
- [x] All 88 existing tests pass (zero regressions)

### CODE REVIEW

**Comment by benthomasson** [comment]
# Code Review Report

**Branch:** benthomasson/ftl-reasons#176
**Models:** claude, gemini
**Gate:** [BLOCK] BLOCK

## Disagreements

The following changes have different verdicts across models:

- **reasons_lib/check_stale.py:pin_lines** [HIGH]: claude: PASS, gemini: BLOCK
- **reasons_lib/check_stale.py:file_lines_changed_since** [LOW]: claude: CONCERN, gemini: PASS

## claude [CONCERN]

### reasons_lib/check_stale.py:parse_diff_hunks
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Clean regex parser for unified diff hunk headers. Correctly handles optional count field (defaults to 1), skips zero-count pure-insertion hunks, and returns old-side ranges as documented. 6 unit tests cover all edge cases including empty input and no-hunk diffs.

### reasons_lib/check_stale.py:file_lines_changed_since
**Verdict:** CONCERN
**Correctness:** QUESTIONABLE
**Spec Compliance:** N/A
**Test Coverage:** PARTIAL
**Integration:** WIRED

The function runs two diffs — `since_sha..HEAD` (committed) and `HEAD..working` (uncommitted). The first diff's old-side line numbers are relative to `since_sha`, which matches the pinned line numbers. However, the second diff's old-side line numbers are relative to HEAD. If committed changes between `since_sha` and HEAD **deleted** lines above the pinned range, the content at the pinned lines shifts downward in HEAD. An uncommitted edit to those shifted lines would have a hunk header at the shifted position (e.g., original line 50 is now at line 45 after 5 deletions), and the overlap check against the original pinned range (50-60) would miss it — a false negative. Since `check_stale` uses this as the sole gate (`continue` on False, skipping content hash), this could cause a stale node to appear fresh. The scenario requires deletions above the pinned range in committed history + uncommitted edits to shifted lines, which is narrow but real. Also, `filepath.name` (not repo-relative path) is used in git commands, which could match wrong files in repos with duplicate filenames — though this is consistent with the existing `file_changed_since`.

### reasons_lib/check_stale.py:check_stale (modified section)
**Verdict:** CONCERN
**Correctness:** QUESTIONABLE
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Two concerns. (1) The `pinned_lines` parsing (`parts = pinned_lines.split("-"); ls, le = int(parts[0]), int(parts[1])`) has no error handling. While `pin_lines` produces well-formed values, a manually edited DB row with malformed metadata (e.g., `"abc"` or `"1-2-3"`) would throw an unhandled exception and abort the staleness check for all remaining nodes. A try/except that falls back to whole-file checking would be more robust. (2) When `file_lines_changed_since` returns False and we `continue`, the SHA auto-bump logic at the bottom of the loop is never reached. Over time, `pinned_sha` falls further behind HEAD, making each diff larger. Not a correctness issue but a performance concern for long-lived pinned beliefs.

### reasons_lib/check_stale.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Input validation is solid — rejects `line_start < 1`, `line_end < line_start`, missing node, and missing source. Auto-pin logic correctly backfills `pinned_sha`, `verified_at`, and `source_hash` when missing. Preserves existing `pinned_sha` if already set. The repos fallback to `network.repos` is defensive but harmless given the API layer also handles it. 6 unit tests cover all paths.

### reasons_lib/api.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** PARTIAL
**Integration:** WIRED

Standard API wrapper following the established pattern (lazy import, `_with_network` context manager, repo path conversion). Uses `write=True` to persist changes. No PG dispatch — intentional, and enforced at the CLI layer via `_require_sqlite`. No direct unit tests for the API wrapper itself, but end-to-end coverage exists through CLI and check_stale tests.

### reasons_lib/cli.py (cmd_pin_lines + cmd_show changes + parser)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** PARTIAL
**Integration:** WIRED

CLI handler properly catches `KeyError`/`ValueError` and exits with appropriate error messages. Parser wiring is correct — `start` and `end` are `type=int`. The `cmd_show` enhancement displays pinned SHA (truncated to 12 chars) and line range when present, consistent with existing output patterns. Dispatch table entry is correctly placed. No CLI-level tests, but this follows the project's existing pattern where CLI commands are tested indirectly through API tests.

### tests/test_pin_sources.py (all new test classes)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

21 new tests across 4 test classes. `TestParseDiffHunks` (6 tests) covers parsing edge cases. `TestFileLinesChangedSince` (5 tests) covers committed/uncommitted/out-of-range/error paths. `TestPinLines` (6 tests) covers happy path, auto-pin, validation, and error cases. `TestCheckStaleLineRange` (4 tests) provides integration coverage for the full check_stale flow with pinned lines. The `test_auto_bumps_sha_preserves_pinned_lines` test asserts pinned_lines survives SHA bumping but doesn't verify the SHA was actually bumped (it captures `new_head` but never asserts against it). Missing: no test for the line-number-drift edge case in uncommitted changes after deletions above the pinned range.

### Self-Review
**Limitations:** Could not inspect the PG backend's `check_stale` implementation to verify whether it would silently ignore `pinned_lines` metadata or error on it. Could not see the full `Node` dataclass definition to verify metadata is always initialized as a dict (relevant to the `node.metadata.get("pinned_lines")` call where metadata could theoretically be None, though the code does check `if node.metadata` first).

### Feature Requests
- Include the full Node dataclass definition when metadata fields are being added/read, so reviewers can verify initialization guarantees
- Show the PG backend's parallel implementation when the SQLite version is modified, to flag feature parity gaps

## gemini [BLOCK]

### reasons_lib/check_stale.py:parse_diff_hunks
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Correctly parses unified diff hunk headers to extract the affected line ranges on the "old" side of the file. The logic properly handles single-line hunks and correctly ignores zero-count hunks (pure insertions), which do not invalidate existing pinned lines.

### reasons_lib/check_stale.py:file_lines_changed_since
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Effectively detects modifications to specific line ranges by checking both committed changes (since the pinned SHA) and uncommitted changes in the working tree. The fail-open strategy (returning `True` on git errors) ensures staleness checks fall back to safe content hashing if git operations fail.

### reasons_lib/check_stale.py:pin_lines
**Verdict:** BLOCK
**Correctness:** BROKEN
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

1. **Metadata Crash**: The function will crash with a `TypeError` if `node.metadata` is `None` (a state explicitly checked for and handled in other parts of `check_stale.py`). Lines 445 and 448 attempt to access `node.metadata` without checking if it exists.
2. **Hash Desync**: When auto-pinning a new SHA, the function does not unconditionally update `node.source_hash` to match the current file state. If the file was already different from the recorded hash before pinning, `check_stale` will subsequently report the node as "Fresh" (because no changes occurred *since* the pin), effectively masking pre-existing staleness.

### reasons_lib/check_stale.py:check_stale
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Seamlessly integrates line-aware checks into the existing staleness loop. It correctly falls back to whole-file checks if no line range is pinned or if `git_aware` is disabled.

### reasons_lib/api.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Standard API wrapper with proper transaction handling (via `_with_network`) and path resolution.

### reasons_lib/cli.py
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Properly exposes the new functionality. The use of `_require_sqlite` in `cmd_pin_lines` is an appropriate safeguard since the Postgres backend currently lacks support for this feature.

### reasons_lib/pg.py
**Verdict:** CONCERN
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** UNTESTED
**Integration:** PARTIAL

The Postgres implementation of `check_stale` in `pg.py` was not updated to handle `pinned_lines`. While the CLI currently blocks setting these pins on Postgres databases, the inconsistency in the core logic library could lead to unexpected behavior if pins are introduced via other means (e.g., belief imports).

### tests/test_pin_sources.py
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Excellent test coverage. It includes unit tests for the regex parser, mocked git repository tests for the diff logic, and full integration tests for the staleness detection flow.

### Self-Review
**Limitations:** I do not have the full definition of the `Node` class, but I inferred that `node.metadata` can be `None` based on frequent guard clauses in the surrounding code (e.g., `if node.metadata`). The lack of a `pg.py` update is flagged as a concern for architectural parity despite the CLI-level block.

### Feature Requests
- The two git commands in `file_lines_changed_since` can be simplified into a single call: `git diff -U0 since_sha -- filename`. This command compares the specified commit directly against the working tree (staged + unstaged), making it both more efficient and more robust.
- Ensure `pin_lines` (and similar pinning functions) always update `node.source_hash` to match the pinned version to prevent "false fresh" reports.

**Comment by benthomasson** [comment]
# Code Review Report

**Branch:** benthomasson/ftl-reasons#176
**Models:** claude, gemini
**Gate:** [CONCERN] CONCERN

## claude [CONCERN]

### reasons_lib/check_stale.py:check_stale (modification)
**Verdict:** CONCERN
**Correctness:** QUESTIONABLE
**Spec Compliance:** N/A
**Test Coverage:** PARTIAL
**Integration:** WIRED

The control flow is correct — malformed `pinned_lines` falls through to content hash, valid lines that changed fall through to content hash, unchanged lines get SHA-bumped and `continue`. However, when pinned lines are unchanged but the rest of the file changed, the code bumps `pinned_sha` and continues without updating `source_hash`. This means a subsequent `check_stale(..., git_aware=False)` will see a stale `source_hash` and falsely flag the node. The test `test_auto_bumps_sha_preserves_pinned_lines` masks this by reverting content back to original. Consider adding `node.source_hash = hash_file(path)` in the auto-bump path (line ~118 in the diff), or at minimum adding a test where content outside the pinned range changes and is NOT reverted to document the expected behavior.

### reasons_lib/check_stale.py:parse_diff_hunks
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Regex correctly parses unified diff `@@ -start,count ... @@` headers. Correctly defaults count to 1 when omitted, skips count=0 (pure insertions), and returns inclusive `(start, start+count-1)` ranges on the old side. Six tests cover the important cases.

### reasons_lib/check_stale.py:file_lines_changed_since
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Uses `git diff -U0 since_sha -- filename` which captures both committed and uncommitted changes in a single call. The overlap check `hunk_end >= line_start and hunk_start <= line_end` is the standard range intersection test. Fails open (returns True) on git errors and timeout, consistent with `file_changed_since`. Uses `filepath.name` with `cwd=filepath.parent`, matching the existing convention in `file_changed_since` and `get_file_commit_sha`. Five tests cover the key scenarios.

### reasons_lib/check_stale.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Input validation is correct (line_start >= 1, line_end >= line_start, node exists, node has source). Auto-pins SHA when missing. Minor redundancy: `node.metadata.get("agent") if node.metadata else None` at line ~455 — `node.metadata` is guaranteed non-None at that point since it was set to `{}` three lines above. Harmless but slightly misleading. Six tests cover the happy path and all error cases.

### reasons_lib/api.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** PARTIAL
**Integration:** WIRED

Follows the same pattern as `pin_update` — loads network with `_with_network(write=True)`, resolves repos, delegates to `check_stale.pin_lines`. No PG dispatch, which is intentional given the CLI's `_require_sqlite` guard. No direct unit test for the API wrapper (tests exercise `check_stale.pin_lines` directly), but the CLI integration test implicitly covers this path.

### reasons_lib/cli.py (pin-lines command)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** UNTESTED
**Integration:** WIRED

Subparser correctly wired with `node_id`, `start` (int), `end` (int). Handler catches `KeyError`/`ValueError`, formats output, exits 1 on error. Dispatch dict includes `"pin-lines": cmd_pin_lines`. No CLI-level test (e.g., subprocess call to `reasons pin-lines`), but the pattern is consistent with other commands that also lack CLI-level tests.

### reasons_lib/cli.py:cmd_show (pinned display)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** UNTESTED
**Integration:** WIRED

Shows `Pinned: <sha12> (lines X-Y)` in `show` output. Truncates SHA to 12 chars, conditionally appends line info. Clean addition. No test for this display change, but it's a display-only enhancement.

### tests/test_pin_sources.py (new test classes)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

21 new tests across 4 classes. `TestParseDiffHunks` (6 tests), `TestFileLinesChangedSince` (5 tests using real git repos), `TestPinLines` (6 tests covering validation and auto-pin), `TestCheckStaleLineRange` (4 integration tests). Tests are well-structured and cover the key scenarios. One gap: no test where content outside pinned lines changes without revert, which would expose the source_hash staleness issue noted above.

### Self-Review
**Limitations:** Could not see the PG backend's `check_stale` implementation (`reasons_lib/pg.py:821`) to verify whether the PG path handles `pinned_lines` metadata or silently ignores it. Could not verify whether `node.metadata` being `None` vs `{}` is consistent across storage round-trips (SQLite load/save).

### Feature Requests
- Include the PG backend's corresponding methods when a SQLite function is modified, to flag divergence
- Show the full test fixture (e.g., `git_repo` conftest) so I can verify test setup assumptions

## gemini [PASS]

### reasons_lib/api.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

The API wrapper correctly implements the interface for pinning lines. It properly manages the network context using `_with_network` and handles repository path resolution. While it is currently limited to the SQLite backend (consistent with existing `pin-update` and `pin-sources` commands), it is correctly wired for that path.

### reasons_lib/check_stale.py
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

The implementation of line-level staleness detection is robust. 
- `parse_diff_hunks` correctly extracts line ranges from the "old" side of a git diff, which is the correct reference for pinned lines.
- `file_lines_changed_since` uses `git diff -U0` efficiently to check for overlaps between the pinned range and actual changes.
- The integration into `check_stale` allows skipping expensive content hashing when the specific lines of interest are untouched, while still providing a fail-safe fallback (falling through to content hash if the line check indicates a change or if git errors occur).
- `pin_lines` handles auto-pinning and source hash updates correctly when establishing a new pin.

### reasons_lib/cli.py
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

The CLI changes correctly expose the new functionality. `pin-lines` is added as a subcommand with appropriate arguments, and the `show` command is updated to display the pinned range. The use of `_require_sqlite` properly enforces the current backend limitation.

### tests/test_pin_sources.py
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Excellent test suite. It covers unit logic for diff parsing (including edge cases like single-line hunks and zero-count insertions), the range overlap logic, and full integration tests with a mocked git environment. The tests verify that changes outside the pinned range correctly result in an auto-bump of the SHA without marking the node as stale.

### Self-Review
**Limitations:** The review confirms that the feature is fully implemented and tested for the SQLite/local-file workflow. I have verified via observations that the `re` and `subprocess` dependencies are present in `check_stale.py`. The lack of Postgres support is noted but appears to be an intentional architectural consistency with other "pinning" features in this project.

**Comment by benthomasson** [comment]
# Code Review Report

**Branch:** benthomasson/ftl-reasons#176
**Models:** claude, gemini
**Gate:** [CONCERN] CONCERN

## Disagreements

The following changes have different verdicts across models:

- **reasons_lib/check_stale.py:pin_lines** [LOW]: claude: CONCERN, gemini: PASS

## claude [CONCERN]

### reasons_lib/check_stale.py:check_stale (modified)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

The branching logic is correct. When `pinned_lines` is set and lines haven't changed: auto-bump SHA + continue (not stale). When lines changed: fall through to content hash comparison (correctly reports stale). When `pinned_lines` is malformed: exception caught, `pass` skips the `else` block, falls through past the `elif` to content hash fallback. The `source_hash` update during auto-bump is intentional — keeps the hash in sync even when changes are outside pinned lines, preventing false stale reports on subsequent runs.

### reasons_lib/check_stale.py:parse_diff_hunks
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Regex correctly parses unified diff hunk headers. Extracts OLD-side ranges, which is the right choice — pinned_lines reference positions at `pinned_sha`, matching the old side. Count-0 exclusion correctly filters pure insertions. Range calculation `(start, start + count - 1)` gives correct inclusive bounds. Six unit tests cover all edge cases including empty diff, no headers, missing count, and zero count.

### reasons_lib/check_stale.py:file_lines_changed_since
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

`git diff -U0 since_sha -- filename` correctly diffs the pinned commit against the working tree (capturing committed + staged + unstaged changes) with zero context lines for exact hunk boundaries. The overlap check `hunk_end >= line_start and hunk_start <= line_end` is the standard range intersection test. Fails open (returns True) on git errors and timeouts, consistent with `file_changed_since`. Uses `filepath.name` with `cwd=filepath.parent`, matching the pattern used by `get_file_commit_sha` and `file_changed_since`.

### reasons_lib/check_stale.py:pin_lines
**Verdict:** CONCERN
**Correctness:** QUESTIONABLE
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Input validation and auto-pin logic are sound. One minor issue: after `if not node.metadata: node.metadata = {}`, the subsequent `agent = node.metadata.get("agent") if node.metadata else None` has a redundant truthiness check — `node.metadata` is guaranteed non-None at that point. Not a bug, just dead code in the conditional. The auto-pin path correctly backfills both `pinned_sha` and `source_hash`, and sets `verified_at`.

### reasons_lib/api.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** PARTIAL
**Integration:** WIRED

Follows the exact pattern of `pin_update()` — same repo resolution, same `_with_network(write=True)` usage. The repo dict conversion (`str` → `Path`) happens here, and `_pin_lines` does a redundant conversion from `network.repos` if repos is None, but the API wrapper already handles that case, so no double-conversion occurs in practice. No PG dispatch, which is consistent with `pin_update` and the CLI's `_require_sqlite` guard. No direct unit test for the API wrapper layer (tests hit `check_stale.pin_lines` directly), but integration via CLI is covered.

### reasons_lib/cli.py:cmd_show (modified)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** UNTESTED
**Integration:** WIRED

Clean display of pinned info. Truncates SHA to 12 chars. Conditionally appends line range. The `result.get("pinned_sha")` falsy check correctly handles both missing key and empty string `""` from the API.

### reasons_lib/cli.py:cmd_pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** UNTESTED
**Integration:** WIRED

Properly uses `_require_sqlite`, catches `KeyError`/`ValueError` with stderr output and exit(1). Argparser correctly uses `type=int` for line numbers. Dispatch table wired. No CLI integration test, but the function is thin enough that unit tests on the underlying `api.pin_lines` and `check_stale.pin_lines` provide adequate coverage.

### tests/test_pin_sources.py (new test classes)
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Comprehensive test suite with 22 new tests across 4 test classes. `TestCheckStaleLineRange` tests the full integration path (check_stale with pinned_lines). `TestPinLines` covers validation, auto-pin, and preservation. `TestFileLinesChangedSince` covers committed, uncommitted, and out-of-range changes. `TestParseDiffHunks` covers parsing edge cases. All tests use the existing `git_repo` fixture and helper functions. The `_multiline_file` helper is duplicated between `TestFileLinesChangedSince` and `TestCheckStaleLineRange` — minor but not worth flagging.

### Self-Review
**Limitations:** Could not see the PG backend's `check_stale` implementation to verify it handles `pinned_lines` metadata gracefully (ignoring it vs. erroring). Could not run the test suite to verify all 22 new tests pass. Did not see the `_require_sqlite` implementation to confirm it properly gates PG usage.

### Feature Requests
- Show the PG backend's parallel implementation when the SQLite version is modified, to flag feature parity gaps
- Include CLI `_require_sqlite` and similar guard functions in observations when CLI commands reference them

## gemini [PASS]

### reasons_lib/api.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Correct implementation of the API wrapper for pinning line ranges. It follows the established pattern of using `_with_network` for atomic operations and handles repository path resolution consistently with other API functions.

### reasons_lib/check_stale.py:check_stale
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Successfully integrates line-aware staleness checks into the main verification loop. The logic to auto-bump the `pinned_sha` when changes occur strictly outside the pinned range is a valuable optimization that keeps metadata fresh without triggering false staleness reports.

### reasons_lib/check_stale.py:parse_diff_hunks
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Robust regex-based parsing of unified diff headers. It correctly extracts OLD-side ranges, handles single-line hunks (where the count is omitted), and appropriately ignores pure insertions (count=0) that don't shift or modify existing lines in the pinned range.

### reasons_lib/check_stale.py:file_lines_changed_since
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Efficiently utilizes `git diff -U0` to detect modifications. The overlap detection logic (`hunk_end >= line_start and hunk_start <= line_end`) is correct for inclusive line ranges. The function fails safe (returns `True`) on Git errors or timeouts, ensuring staleness is re-verified via content hashing.

### reasons_lib/check_stale.py:pin_lines
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Solid implementation for persisting line metadata. The inclusion of input validation and the auto-pinning of the current Git SHA (when missing) makes the feature robust and user-friendly.

### reasons_lib/cli.py
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Clean integration into the CLI. The updates to the `show` command provide clear visibility into pinned lines, and the new `pin-lines` subcommand follows standard argument patterns.

### tests/test_pin_sources.py
**Verdict:** PASS
**Correctness:** VALID
**Spec Compliance:** N/A
**Test Coverage:** COVERED
**Integration:** WIRED

Comprehensive test suite that covers both low-level logic (diff parsing, range overlap) and high-level integration (CLI behavior, `check_stale` workflows). The use of a real Git repository fixture for these tests ensures high confidence in the implementation.

### Self-Review
**Limitations:** None. All relevant function bodies and test files were provided and reviewed.

### Feature Requests
- The logic for Git interactions (parsing diffs, running commands) is currently embedded in `check_stale.py`. As the project's Git-awareness grows, moving these into a dedicated `git_utils.py` or similar would improve maintainability.
